dotnet/src/SemanticKernel.Abstractions/SkillDefinition/FunctionsView.cs

```diff
@@ -94,22 +77,120 @@ public bool IsSemantic(string skillName, string functionName)
+
+            return result;
+        }
+
+        public IEnumerator<KeyValuePair<string, IReadOnlyCollection<FunctionView>>> GetEnumerator()
+        {
+            return new ReadOnlyFunctionsViewEnumerator(this._decorated.GetEnumerator());
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return new ReadOnlyFunctionsViewEnumerator(this._decorated.GetEnumerator());
+        }
+
+        public bool ContainsKey(string key)
+        {
+            return this._decorated.ContainsKey(key);
+        }
+
+        private IEnumerable<IReadOnlyCollection<FunctionView>> GetValues()
+        {
+            foreach (var decoratorValue in this._decorated.Values)
+            {
+                yield return decoratorValue.ToArray();
+            }
+        }
+    }
+
+    private class ReadOnlyFunctionsViewEnumerator : IEnumerator<KeyValuePair<string, IReadOnlyCollection<FunctionView>>>
+    {
+        private readonly IEnumerator<KeyValuePair<string, List<FunctionView>>> _decorated;
+
+        public ReadOnlyFunctionsViewEnumerator(IEnumerator<KeyValuePair<string, List<FunctionView>>> decorated)
+        {
+            this._decorated = decorated;
+        }
+
+        public KeyValuePair<string, IReadOnlyCollection<FunctionView>> Current => new(this._decorated.Current.Key, this._decorated.Current.Value.ToArray());
+
+        object IEnumerator.Current => this.Current;
+
+        public void Dispose()
+        {
+            this._decorated.Dispose();
+        }
+
+        public bool MoveNext()
+        {
+            return this._decorated.MoveNext();
+        }
+
+        public void Reset()
+        {
+            this._decorated.Reset();
+        }
+    }
+
+    private bool IsFunctionCheck(string skillName, string functionName, bool nativeFunctionCheck)
+    {
+        this._semanticFunctions.TryGetValue(skillName, out var semanticFunctions);
+        var foundSemanticFunction = semanticFunctions?.Any(x => string.Equals(x.Name, functionName, StringComparison.OrdinalIgnoreCase))
+                                    ?? false;
 
-        var nf = this.NativeFunctions.ContainsKey(skillName)
-                 && this.NativeFunctions[skillName]
-                     .Any(x => string.Equals(x.Name, functionName, StringComparison.OrdinalIgnoreCase));
+        this._nativeFunctions.TryGetValue(skillName, out var nativeFunctions);
+        var foundNativeFunction = nativeFunctions?.Any(x => string.Equals(x.Name, functionName, StringComparison.OrdinalIgnoreCase))
+                                  ?? false;
 
-        if (sf && nf)
+        if (foundSemanticFunction && foundNativeFunction)
         {
             throw new AmbiguousMatchException("There are 2 functions with the same name, one native and one semantic");
         }
 
-        return nf;
+        return (nativeFunctionCheck)
+            ? foundNativeFunction
+            : foundSemanticFunction;
```

Comment:
nit: I think this can be one-liner without parentheses:
```suggestion
        return nativeFunctionCheck ? foundNativeFunction : foundSemanticFunction;
```